include/iso9660.h

```diff
@@ -0,0 +1,232 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+ * USA.
+ */
+
+/**
+ * Collection of utility functions that are deticated specifically to reading
+ * an ISO 9660 image.
+ */
+
+#ifndef ISO9660_ISO9660_H_
+#define ISO9660_ISO9660_H_
+
+#include <cstdint>
+#include <cstdlib>
+
+#include <array>
+#include <fstream>
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace iso9660 {
+
+enum class SectorType {
+  BOOT_RECORD = 0,
+  PRIMARY = 1,
+  SUPPLEMENTARY = 2,
+  PARTITION = 3,
+  SET_TERMINATOR = 255
+};
+
+/**
+ * ECMA-119 calls this a directory record but it's actually either a file or a
+ * directory so it's a file.
+ */
+struct File {
+  enum class Flag {
+    /*
+     * Set if file is hidden.
+     */
+    HIDDEN = 1,
+    /*
+     * Set if entry is a directory.
+     */
+    DIRECTORY = 1 << 1,
+    /*
+     * Set if entry is an associated file.
+     */
+    ASSOCIATED = 1 << 2,
+    /*
+     * Set if information is structured according to the extended attribute
+     * record.
+     */
+    EXTENDED_STRUCTURE = 1 << 3,
+    /*
+     * Set if owner, group and permissions are specified in the extended
+     * attribute.
+     */
+    EXTENDED_PERMISSIONS = 1 << 4,
+    /*
+     * Set if file has more than one directory record.
+     */
+    MULTIPLE_RECORDS = 1 << 7
+  };
+
+  std::size_t length;
+  std::size_t extended_length;
+  std::size_t location;
+  std::size_t size;
+  int datetime;
+  int flags;
+  int file_unit_size;
+  int interleave_gap_size;
+  int volume_sequence_number;
+  std::string name;
+
+  bool has(Flag flag) const;
+  bool isdir() const;
+  std::size_t max_growth() const;
+};
+
+/**
+ * ECMA-119 calls this a path table record which is basically a directory.
+ */
+struct Directory {
+  std::size_t size;
+  std::size_t extended_length;
+  std::size_t location;
+  int parent;
+  std::string name;
+  std::vector<File> files;
+};
+
+struct VolumeDescriptorHeader {
+  SectorType type;
+  /*
+   * Identifier specifies according to what specification the following data
+   * was recorded.  The following information was taken from ECMA-167/3:
+   * BEA01: 2/9.2 Beginning extended area descriptor.
+   * BOOT2: 2/9.4 Boot descriptor.
+   * CD001: ECMA-119
+   * CDW02: ECMA-168
+   * NSR02: ECMA-167/2
+   * NSR03: ECMA-167/3
+   * TEA01: 2/9.3 Terminating extended area descriptor.
+   */
+  std::string identifier;
+  int version;
+};
+
+/**
+ * The "generic" volume descriptor for the primary and supplementary volume
+ * descriptor.
+ */
+struct VolumeDescriptor {
```

Comment:
I don't particularly like the overall structure of this file. I guess a bit more of object-oriented design would help with the readability a lot.
For example this structure should (at least in my opinion) be a class and the function that creates it should be its constructor. More complicated operations on it then can be implemented as its methods.
Same goes for the structs above, too.